### ISSUE TITLE
Unable to get access token for tenant

### ISSUE THREAD

**Issue by test10keystep** [Yvand/AzureCP#48]
The installation is related to a test environment prior to implementing the AzureCP in the production environment.

While configuring AzureCP in accordance with the instructions, the following error pops up while testing the connection to the tenant ( the name, AppID and App secret are correct)

The error message:
"Unable to get access token for tenant '**********.onmicrosoft.com': One or more errors occurred".

Using ULS real time, we seem to be getting the following:
[] Unexpected error occurred while getting access token for tenant '**********.onMicrosoft.com': System.Net.WebException: The remote server returned an error: (407) Proxy Authentication Required., Callstack:   
 at System.Net.HttpWebRequest.EndGetResponse(IAsyncResult asyncResult)    
 at System.Net.Http.HttpClientHandler.GetResponseCallback(IAsyncResult ar)

There is no proxy that stands between the test platform and the internet.
Any ideas as to what this could be?
Any tips are greatly appreciated.

Regards
TK

**Comment by Yvand** [comment]
@test10keystep can you open a Windows session on the central admin server, as the SharePoint farm account (which runs the central admin process), and double check that there is no proxy configured?

**Comment by test10keystep** [comment]
I have checked in "internet options" and even the winHTTP proxy using netsh. No proxy is configured. I talked to the network guys and explicitly asked them to turn off proxy settings for this server completely (not just proxy authentication). 
In addition I checked that all servers have access to the internet on port 443 out.
All network scans show the SharePoint server attempting a connection to 13.107.5.80:443 but I still recieve the 407 proxy authentication required in the log as it fails.
The local config file has 
<system.net>
    <defaultProxy />
</system.net>

Still no where closer to a solution.

**Comment by test10keystep** [comment]
It looks like adding this **defaultProxy useDefaultCredentials="true"**  to all config files in question on all farm servers allows for a successful test of the tenant connection and the test is accesses the tenant token.

**Comment by test10keystep** [comment]
The remaining question as a result of the AzureCP, is it possible to delegate permissions to users in bulk after AzureCP is in place using PowerShell? 
The GUI seems to process delation requests, while in PS the error "The specified user could not be found" shows up.
Any comment is welcomed as I feel we are almost reaching our set goals and would hate to see a reversal on the AzureCP by management.

**Comment by Yvand** [comment]
@test10keystep Thanks for posting how you fixed the proxy issue.
I'm not sure I get your question. What do you mean by delegating permissions to users in bulk? You can grant permissions with PowerShell using [New-SPUser](https://docs.microsoft.com/en-us/powershell/module/sharepoint-server/new-spuser?view=sharepoint-ps), which will work just fine with AzureCP

**Comment by test10keystep** [comment]
Using the New-SPuser did not work for me, I receive the following error message "The specified user externaluser@external_domain.com could not be found"

After attempting a multitude of scenarios, we were capable of importing external users in bulk into groups using Add-PnPUserToGroup which seems to validate across AzureCP in PowerShell.

**Comment by Yvand** [comment]
Your scenario is not clear to me. Those users you can't add with New-SPUser, are they Azure AD users?
If so, can you check the SharePoint logs and filter on Product/Area AzureCP to see why AzureCP couldn't find them?

**Comment by test10keystep** [comment]
The users in question here are Azure invited guest users.

**Comment by test10keystep** [comment]
When using the Add-PnPUserToGroup it works.
When I utilize the New-SPuser I receive the following 3 errors:
► [AzureCP] Unexpected error(s) occurred while querying Azure AD tenant 'Tenant.onMicrosoft.com':  
[EXCEPTION 1]: System.Net.Http.HttpRequestException: An error occurred while sending the request.. Callstack:   
 at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()    
 at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)    
 at Microsoft.Graph.HttpProvider.<SendRequestAsync>d__21.MoveNext()

► [AzureCP] Got 0 users/groups in 10 ms from 'Tenant.onMicrosoft.com' with input 'testUser@domain.com'

► [AzureCP] Validation failed: found 0 entities instead of 1 for incoming claim with value 'testUser@domain.com' and type 'http://schemas.xmlsoap.org/ws/2005/05/identity/claims/emailaddress'

**Comment by Yvand** [comment]
AzureCP works with Azure AD Guest users works, but there is a design issue in Azure AD which makes it difficult to handle, you can find more information here: https://feedback.azure.com/forums/169401-azure-active-directory/suggestions/36274663-sign-in-to-sharepoint-2016-with-guest-accounts-is

About the error in the logs, it's not normal, do you reproduce it each time or is it random?

**Comment by test10keystep** [comment]
Hello. thank you for the information, I am already familiar with Azure's #EXT# issue.
The majority of our invited external guest users are capable of authenticating with the constructed claim string without the #EXT#.

As for your question regarding the errors, yes, this is something that can be regenerated and happens each time we use the new-spuser cmdlet and for all guest accounts regardless of the guest account being with or without a tenancy.

As I mentioned earlier, we do not run into this problem if the PnP cmdlet "Add-PnPUserToGroup" is used.

We can solve certain issues with joining guest accounts to SharePoint groups with the Add-PnPUserToGroup, but not all.
As the situation is today within the test environment, I cannot recommend the utilization of AzureCP in the production environment until all creases are ironed out for us.
I  wish to also mention that I greatly appreciate your efforts.

**Comment by Yvand** [comment]
@test10keystep since it's not random I should be able to reproduce this issue.
Can you please describe the exact steps I need to follow to repro it?

**Comment by test10keystep** [comment]
It was a very basic standard installation of AzureCP as described in the instructional information on the web.
The actually configuration was troublesome from a proxy persepective as previously documented earlier in my initial regestration of the issue. That was resolved as you know by securing that proxy was turned off and that the .config file edited to contain **defaultProxy useDefaultCredentials="true"** , this resolved the initial problem in regards to accessing the tenant token while testing the connection.

Shortly after, our attempts at automating failed when using **New-SPuser** resulting in the error messages listed above  in my previous post, simultaneously using the people picker from within the GUI functioned every single time and was successful.

PnP is implemented to help with complex provisioning and management of routines.
When connect-PnPonline cmdlet is executed it is done locally and **NOT** by using the "**UseWebLogin**" switch.

**Connect-PnPOnline -Url https://SharePoint.url.com -Credentials $cred
Add-PnPUserToGroup -LoginName "i:05.t|BlaBlaBla|external_user@external_domain.com" -Identity "SharePointGroup01"**

When **Add-PnPUserToGroup** cmdlet is used, the Azure guest user can be imidiately added to a group as long as the invited user object resides in Azure even if the invited user has not accepted the invitation yet and the "**UserState**" is still "**PendingAcceptance**".

So in conclusion your proposal to using New-SPuser fails but Add-PnPUserToGroup functions flawlessly.
This as a work around is OK, but as a stable solution to recommend for the production environment it is not. 
We cannot explain the sporadic behaviour as long as we have not located the reason behind the error.

What does Add-PnPUserToGroup do differently than New-SPuser in its attempt at locating the user object within Azure via AzureCP ?

I hope this reflects certain light on the setup/situation at hand.

**Comment by Yvand** [comment]
I tested both cmdlets in a new site collection and both can successfully add guest accounts:

```powershell
Add-PnPUserToGroup -LoginName "i:0e.t|yvandev|GuestAccount2@GuestDomain.com" -Identity "temp Members"
New-SPUser -UserAlias "i:0e.t|yvandev|GuestAccount1@GuestDomain.com" -DisplayName "GuestAccount1" -Web "http://spsites/sites/temp/"
```

For both I see that AzureCP validated them successfully:

```
02/28/2019 17:14:09.84	powershell_ise.exe (0x1438)	0x1C14	AzureCP	Claims Picking	1337	High	[AzureCP] Validated entity: display text: 'GuestAccount1@GuestDomain.com', claim value: 'GuestAccount1@GuestDomain.com', claim type: 'http://schemas.xmlsoap.org/ws/2005/05/identity/claims/upn'	5e03cf16-aa35-4a24-8b19-880eae64f905
```

Can you get verbose logs on Product/Area AzureCP when you run New-SPUser?
Are you using the default configuration of AzureCP?

**Comment by test10keystep** [comment]
Yes, the we are using the default configuration of AzureCP.
This is the entry referencing the transaction from a verbose log:

Could not find idenitty for input. Input: 'i:05.t|blablabla|user@domain.com', 
Exception: 'System.Security.Principal.IdentityNotMappedException: 
Some or all identity references could not be translated.     
at System.Security.Principal.NTAccount.Translate(IdentityReferenceCollection sourceAccounts, Type targetType, Boolean forceSuccess)     
at System.Security.Principal.NTAccount.Translate(Type targetType)     
at Microsoft.SharePoint.Utilities.SPUtility.ResolvePrincipalInternalV2(SPWeb web, 
SPWebApplication webApp, Nullable`1 urlZone, String input, SPPrincipalEntityType types, 
SPPrincipalSource sources, SPUserCollection usersContainer, Boolean inputIsEmailOnly, Boolean alwaysAddWindowsResolver)'.	36def50f-5253-2104-fd1e-d0ffee5712c2

**Comment by Yvand** [comment]
@test10keystep You only copied a very little fragment of the error so I'm missing much of the context, but it doesn't seem to be related to AzureCP: SharePoint tries to resolve user as a Windows user.

**Comment by Yvand** [comment]
@test10keystep do you have updates on this issue?

**Comment by test10keystep** [comment]
Sorry for a delayed response.
The implementation of AzureCP has been arrested for the time being.
The results have been unstable for us to proceed into production. 
I thank you for your assistance. I will give feedback if we ever decide to try AzureCP again.

**Comment by Srinivasnamala2003** [comment]
Hi Yvand
People picker is not working when take the logs we are seeing below error in the logs please give me the suggestion to resolving the issue 
we are configure UPA AZure CCP 
Error :
[AzureCP] Unexpected error occurred Microsoft.Graph could not query tenant 'EYGS.onmicrosoft.com': System.Net.Http.HttpRequestException: An error occurred while sending the request., Callstack:   
at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()    
at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)    
at Microsoft.Graph.RedirectHandler.<SendAsync>d__6.MoveNext()  --- End of stack trace from previous location where exception was thrown ---    
at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()    
at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)    
at Microsoft.Graph.RetryHandler.<SendAsync>d__9.MoveNext()  --- End of stack trace from previous location where exception was thrown ---    
at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()    
at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)    
at Microsoft.Graph.CompressionHandler.<SendAsync>d__2.MoveNext()  --- End of stack trace from previous location where exception was thrown ---    
at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()    
at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)    
at Microsoft.Graph.AuthenticationHandler.<SendAsync>d__16.MoveNext()  --- End of stack trace from previous location where exception was thrown ---    
at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()    
at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)    
at Microsoft.Graph.HttpProvider.<SendRequestAsync>d__19.MoveNext()
